File: ibkr_show_backend/app/schemas/dividends.py
from pydantic import BaseModel

from app.schemas.common import PaginationInfo


class DividendItem(BaseModel):
    account_id: str
    currency: str | None = None
    symbol: str | None = None
    description: str | None = None
    date_time: str | None = None
    settle_date: str | None = None
    amount: float | None = None
    gross_amount: float | None = None
    tax: float | None = None
    fx_rate_to_base: float | None = None
    amount_in_base: float | None = None
    flow_direction: str | None = None
    flow_type: str | None = None
    dividend_type: str | None = None
    transaction_id: str | None = None
    report_date: str | None = None
    activity_code: str | None = None
    activity_description: str | None = None
    level_of_detail: str | None = None


class DividendListResponse(BaseModel):
    items: list[DividendItem]
    pagination: PaginationInfo


class DividendSummaryResponse(BaseModel):
    record_count: int
    total_amount: float | None = None
    total_gross_amount: float | None = None
    by_symbol: dict[str, float]

File: ibkr_show_frontend/src/api/dividends.ts
import { request } from './http'
import type { PaginationInfo } from '@/types/common'

export interface DividendItem {
  account_id: string
  currency: string | null
  symbol: string | null
  description: string | null
  date_time: string | null
  settle_date: string | null
  amount: number | null
  gross_amount: number | null
  tax: number | null
  fx_rate_to_base: number | null
  amount_in_base: number | null
  flow_direction: string | null
  flow_type: string | null
  dividend_type: string | null
  transaction_id: string | null
  report_date: string | null
  activity_code: string | null
  activity_description: string | null
  level_of_detail: string | null
}

export interface DividendListResponse {
  items: DividendItem[]
  pagination: PaginationInfo
}

export interface DividendSummaryResponse {
  record_count: number
  total_amount: number | null
  total_gross_amount: number | null
  by_symbol: Record<string, number>
}

export interface DividendQuery {
  start_date?: string
  end_date?: string
  symbol?: string
  sort_by?: string
  sort_order?: 'asc' | 'desc'
  page?: number
  page_size?: number
}

function toQueryString(params: Record<string, string | number | undefined | null> | DividendQuery): string {
  const searchParams = new URLSearchParams()
  Object.entries(params).forEach(([key, value]) => {
    if (value !== undefined && value !== '') {
      searchParams.set(key, String(value))
    }
  })
  const queryString = searchParams.toString()
  return queryString ? `?${queryString}` : ''
}

export function fetchDividends(params: DividendQuery): Promise<DividendListResponse> {
  return request<DividendListResponse>(`/api/dividends${toQueryString(params)}`)
}

export function fetchDividendSummary(params: Omit<DividendQuery, 'sort_by' | 'sort_order' | 'page' | 'page_size'>): Promise<DividendSummaryResponse> {
  return request<DividendSummaryResponse>(`/api/dividends/summary${toQueryString(params)}`)
}